# データサイエンス特論 第11回 演習課題
青山学院大学大学院 理工学研究科 理工学専攻 知能情報コース 修士1年 35626302 森下剛

In [42]:
import statsmodels.api as sm
from sklearn.datasets import load_diabetes
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, root_mean_squared_error
from sklearn.model_selection import cross_val_score

## 回帰分析

#### 演習問題1の回答

In [43]:
diabetes_data = load_diabetes()

diabetes_df = pd.DataFrame(diabetes_data.data, columns=diabetes_data.feature_names)
diabetes_df["target"] = diabetes_data.target

X = diabetes_df[diabetes_data.feature_names]
y = diabetes_df["target"]

X = sm.add_constant(X)

# 重回帰分析
model = sm.OLS(y, X)
result = model.fit()

print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                 target   R-squared:                       0.518
Model:                            OLS   Adj. R-squared:                  0.507
Method:                 Least Squares   F-statistic:                     46.27
Date:                Tue, 23 Jun 2026   Prob (F-statistic):           3.83e-62
Time:                        16:23:00   Log-Likelihood:                -2386.0
No. Observations:                 442   AIC:                             4794.
Df Residuals:                     431   BIC:                             4839.
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        152.1335      2.576     59.061      0.0

#### 演習問題2の回答

##### 上段
- Dep. Variable：目的変数の名前
- Model：使用した推定モデルの種類
- Method：パラメータの推定方法
- No. Observations：分析に使用したサンプル数
- Df Residuals：残差の自由度
- Df Model：モデルの自由度
- R-squared：決定係数
- Adj. R-squared：自由度調整済み決定係数
- F-statistic：すべての回帰係数が同時に0かどうかを検定する統計量
- Prob (F-statistic)：F検定のp値（小さいほどモデル全体が統計的に有意）
- Log-Likelihood：対数尤度（モデルのデータへの当てはまりの度合い）
- AIC：赤池情報量規準（小さいほど良い）
- BIC：ベイズ情報量規準（小さいほど良い）

##### 中段
- coef：回帰係数の推定値
- std err：回帰係数の標準誤差
- t：t値
- P>|t|：t検定のp値
- [0.025, 0.975]：回帰係数の95%信頼区間

##### 下段
- Omnibus：残差が正規分布に従うかを調べる検定統計量
- Prob(Omnibus)：Omnibus検定のp値
- Skew：残差の歪度
- Kurtosis：残差の尖度
- Durbin-Watson：残差の自己相関を調べる指標
- Jarque-Bera (JB)：歪度と尖度から残差の正規性を検定する統計量
- Prob(JB)：JB検定のp値（大きいほど正規性が支持される）
- Cond. No.：説明変数間の多重共線性の強さの目安（大きいと共線性が疑われる）

#### 演習問題3の回答

In [44]:
X = diabetes_df[diabetes_data.feature_names]
y = diabetes_df["target"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=1
)

model = LinearRegression()
model.fit(X_train, y_train)

y_train_pred = model.predict(X_train)
y_test_pred = model.predict(X_test)

# MAE
mae_train = mean_absolute_error(y_train, y_train_pred)
mae_test = mean_absolute_error(y_test, y_test_pred)

# RMSE
rmse_train = root_mean_squared_error(y_train, y_train_pred)
rmse_test = root_mean_squared_error(y_test, y_test_pred)

print(f"訓練データ MAE : {mae_train}")
print(f"テストデータ MAE : {mae_test}")
print(f"訓練データ RMSE: {rmse_train}")
print(f"テストデータ RMSE: {rmse_test}")


訓練データ MAE : 44.131896061172775
テストデータ MAE : 41.64919844144016
訓練データ RMSE: 53.740437087615184
テストデータ RMSE: 53.17033023617667


#### 演習問題4の回答

In [45]:
X = diabetes_df[diabetes_data.feature_names]
y = diabetes_df["target"]

model = LinearRegression()

mae_scores = -cross_val_score(model, X, y, cv=5, scoring="neg_mean_absolute_error")
rmse_scores = -cross_val_score(model, X, y, cv=5, scoring="neg_root_mean_squared_error")

for i, (mae, rmse) in enumerate(zip(mae_scores, rmse_scores), start=1):
    print(f"Fold {i}: MAE = {mae}, RMSE = {rmse}")

print(f"MAE の平均: {mae_scores.mean()}")
print(f"RMSE の平均: {rmse_scores.mean()}")


Fold 1: MAE = 43.02616605962198, RMSE = 52.72497936663119
Fold 2: MAE = 44.800480102243256, RMSE = 55.034864757066416
Fold 3: MAE = 48.15571020337361, RMSE = 56.900681785933465
Fold 4: MAE = 43.01303220252327, RMSE = 54.85204179282562
Fold 5: MAE = 42.387107598312724, RMSE = 53.94638716133296
MAE の平均: 44.27649923321497
RMSE の平均: 54.69179097275793
